File: Python/Algorithms/Searching/BinarySearch.py
# Binary Search function
def binary_search(arr, x):
    low = 0
    high = len(arr) - 1
    mid = 0
 
    while low <= high:
        mid = (high + low) // 2
 
        if arr[mid] < x:
            low = mid + 1
 
        elif arr[mid] > x:
            high = mid - 1
 
        else:
            return mid
 
    return -1

# Main function
def main():
    arr = []
    n = int(input("Enter the number of elements: "))
    print("Enter the elements: ")
    for i in range(0, n):
        x = int(input())
        arr.append(x)

    x = int(input("Enter the number to search: "))

    # Function call
    result = binary_search(arr, x)
    
    if result != -1:
        print("Element is present at index", str(result))
    else:
        print("Element is not present in array")

# Main function execution
if __name__ == "__main__":
    main()

''' 
  Practice Problems
  https://www.hackerearth.com/practice/algorithms/searching/binary-search/practice-problems/
'''

File: C++/Algorithms/Searching/Binary Search.cpp
//Algorithm : Binary Search
//Time Complexity : O(log(n)) Regressive Growth rate function
/*
Suitable for Larger array sizes
*/

#include<bits/stdc++.h>
using namespace std;
int main()
{
	ios_base::sync_with_stdio(0); cin.tie(0); cout.tie(0);
	int n; // size of the array
	cin >> n;
	int a[n];
	for (int i = 0; i < n; i++)
	{
		cin >> a[i];
	}
	int val;
	int low = 0;
	int high = n - 1;
	int pos = -1;
	//The value you want to search
	cin >> val;
	while (low <= high)
	{
		int mid = (low + high) / 2;
		if (a[mid] == val)
		{
			pos = mid + 1;
			break;
		}
		else if (a[mid] < val)
		{
			low = mid + 1; // vsl lies in the 2nd-half
		}
		else
		{
			high = mid - 1; //val lies in the 1st-half
		}
	}
	if (pos == -1)
	{
		printf("The element is not present \n");
	}
	else
	{
		printf("The value is present in the %d position", pos);
	}

	return 0;
}

//Problems across coding platforms
//https://codeforces.com/problemset/problem/1566/A
//https://codeforces.com/problemset/problem/1476/A